# Import Libraries

In [2]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

# Import Data

In [4]:
# Create Path
path = r'C:\Users\jenni\OneDrive\Documenten\Data Analyst Program\Data Analytics Immersion\Achievement 4\02-2024 Instacart Basket Analysis'

In [5]:
# Importing Orders Dataframe
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [6]:
# Importing Products Dataframe
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# Exercise: Data Consistency Checks

In [15]:
# Create a Dataframe
df_test = pd.DataFrame()

In [16]:
# Create a Mixed Type Column
df_test['mix'] = ['a', 'b', '1', True]

In [17]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [18]:
#Check for Mixed Types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\jenni\AppData\Local\Temp\ipykernel_15620\675945478.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [19]:
df_test['mix'] = df_test['mix'].astype('str')

In [20]:
# Looking for Missing Values
df_prods.isnull().sum()

Unnamed: 0       0
product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

In [21]:
# Create Dataframe with Missing Values from Product_Name
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [22]:
df_nan

Empty DataFrame
Columns: [Unnamed: 0, product_id, product_name, aisle_id, department_id, prices]
Index: []

In [23]:
df_prods.shape

(49672, 6)

In [24]:
# Create Dataframe without Missing Values in Product_name 
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [25]:
df_prods_clean.shape

(49672, 6)

In [26]:
# Look for Duplicates in Products
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [27]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, product_id, product_name, aisle_id, department_id, prices]
Index: []

In [28]:
# Remove Duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [29]:
df_prods_clean_no_dups.shape

(49672, 6)

In [30]:
# Export products Dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task: Data Consistency Checks

In [31]:
df_ords.describe().round(0)

Unnamed: 0   order_id    user_id  order_number  orders_day_of_week  \
count   3421083.0  3421083.0  3421083.0     3421083.0           3421083.0   
mean    1710541.0  1710542.0   102978.0          17.0                 3.0   
std      987582.0   987582.0    59534.0          18.0                 2.0   
min           0.0        1.0        1.0           1.0                 0.0   
25%      855270.0   855272.0    51394.0           5.0                 1.0   
50%     1710541.0  1710542.0   102689.0          11.0                 3.0   
75%     2565812.0  2565812.0   154385.0          23.0                 5.0   
max     3421082.0  3421083.0   206209.0         100.0                 6.0   

       orders_hour_of_day  days_since_prior_order  
count           3421083.0               3214874.0  
mean                 13.0                    11.0  
std                   4.0                     9.0  
min                   0.0                     0.0  
25%                  10.0                     4.0  
50%                  13.0                     7.0  
75%                  16.0                    15.0  
max                  23.0                    30.0

All time-related columns seem to be in order. The maximum order number is 100, while the count is 3421083, so I would have expected the maximum to be higher

In [32]:
#Check for Mixed Types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [33]:
# Looking for Missing Values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
orders_hour_of_day             0
days_since_prior_order    206209
dtype: int64

## Missing Values Findings
The Days Since Prior Order column is the only one missing values (which can also be seen in the statistical table above). it could be a fault in the system that did not record the date of the last order or such

In [34]:
# Create Dataframe with missing values
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [35]:
df_nan_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         orders_hour_of_day  days_since_prior_order  
0                         8                     NaN  
11                       11                     NaN  
26                       14                     NaN  
39                       11                     NaN  
45                       12                     NaN  
...                     ...                     ...  
3420930                  12                     NaN  
3420934                  18                     NaN  
3421002                  19                     NaN  
3421019                  15                     NaN  
3421069                  11                     NaN  

[206209 rows x 7 columns]

In [38]:
# Imputing Missing Values of Days Since Prior Order
df_ords['days_since_prior_order'].fillna(11, inplace=True)

In [40]:
df_ords.shape

(3421083, 7)

## Addressing Missing Values
I imputed the values with the mean (no outliers) as the column contained integers and the amount of missing values was 6,4% of the total values. So I preferred to use a method that kept the other data while ensuring my analysis will not be off

In [41]:
# Look for Duplicates in Orders
df_dups_ords = df_ords[df_ords.duplicated()]

In [42]:
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, orders_hour_of_day, days_since_prior_order]
Index: []

## Duplicates Findings
There were no duplicates to be found in the Orders dataframe so no need for changes

In [43]:
# Export products Dataframe
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))